File: edu/harvard/cs50/week2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>

// Coleman-Liau index - a formula for determining grade level
//
// index = 0.0588 * L - 0.296 * S - 15.8
//
// L = average number of letters per 100 words in the text.
// S = average number of sentences per 100 words of text.

typedef struct
{
    int letters;
    int words;
    int sentences;
} WordCounts;

WordCounts count(char *text)
{
    int sentcount = 0;
    int lettercount = 0;
    int wordcount = 0;
    int len = strlen(text);
    for (int i = 0; i < len; i++)
    {
        char cur = text[i];
        if ((cur == '.' || cur == '?' || cur == '!') &&
            (isspace(text[i + 1])))
        {
            sentcount++;
        }
        else if (isspace(cur))
        {
            wordcount++;
        }
        else
        {
            lettercount++;
        }
    }

    // complete hack - add one for the last word / sentence
    sentcount++;
    wordcount++;

    WordCounts result;
    result.letters = lettercount;
    result.sentences = sentcount;
    result.words = wordcount;
    return result;
}

int main(int argc, char *argv[])
{
    char *sent = get_string("Text: ");
    WordCounts counts = count(sent);

    float S = ((float)counts.sentences / (float)counts.words) * 100.0;
    float L = ((float)counts.letters / (float)counts.words) * 100.0;
    int index = (int)((0.0588 * L) - (0.296 * S) - 15.8);

    // printf("L = %f S = %f index=%i\n", L, S, index);

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

File: edu/harvard/cs50/week6/sentimental-readability/readability.py
def grade_level(text: str) -> tuple[int, int, int]:
    """Computes the estimated grade level of `text`. using the Coleman-Liau
    index.

    grade level = 0.0588 * L - 0.296 * S - 15.8

    * L = average number of letters per 100 words
    * S = average number of sentences per 100 words

    Your program should count the number of letters, words, and sentences in the
    text.

    * A letter is any upper case or lower case.
    * Any sequence of characters separated by spaces should count as a word.
    * Any occurrence of a period, exclamation point, or question mark indicates
      the end of a sentence.
    """

    letters = 0
    words = 0
    sentences = 0

    for char in text:
        if char in [".", "?", "!"]:
            sentences += 1
        elif char.isspace():
            words += 1
        elif char.isalpha():
            letters += 1

    # hack: add one for the last word
    words += 1
    return (letters, words, sentences)


if __name__ == "__main__":
    letters, words, sentences = grade_level(input("Text: "))

    S = (sentences / words) * 100
    L = (letters / words) * 100

    cl_index = round((0.0588 * L) - (0.296 * S) - 15.8)

    if cl_index < 1:
        print("Before Grade 1")
    elif cl_index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {cl_index}")
